### PR TITLE
Solving #2: Exchange rate available for cryptocurrency.

### PR DESCRIPTION
**Pull request by IoriOikawa**
Rate exchange enabled with main-stream cryptocurrencies.

### CODE REVIEW

File: AccountingServer.sln.DotSettings [hunk 1]
```diff
@@ -152,6 +152,7 @@
 	<s:Boolean x:Key="/Default/UserDictionary/Words/=aggr/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/UserDictionary/Words/=amort/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/UserDictionary/Words/=amorts/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/UserDictionary/Words/=coinmartketcap/@EntryIndexedValue">True</s:Boolean>
```
Comment (b1f6c1c4) [inline AccountingServer.sln.DotSettings]: remove this line

File: AccountingServer.BLL/Util/Exchange.cs [hunk 2]
```diff
@@ -150,4 +191,112 @@ private double Invoke(DateTime date, string from, string to)
...
+            }
+            finally
+            {
+                m_Lock.ExitWriteLock();
+            }
+
+            return rate.Value;
+        }
+    }
+}
```
Comment (b1f6c1c4) [inline AccountingServer.BLL/Util/Exchange.cs]: add new line

File: AccountingServer.BLL/Util/Exchange.cs [hunk 3]
```diff
@@ -150,4 +191,112 @@ private double Invoke(DateTime date, string from, string to)
...
+                var res = req.GetResponse();
+                using (var stream = res.GetResponseStream())
+                {
+                    var reader = new StreamReader(stream ?? throw new NetworkInformationException());
+                    var json = JObject.Parse(reader.ReadToEnd());
+                    if (json["status"]["error_code"].Value<int>() == 0)
+                        rate = json["data"][to]["quote"][from]["price"].Value<double>();
+                }
+            }
+            if (rate.HasValue)
```
Comment (b1f6c1c4) [inline AccountingServer.BLL/Util/Exchange.cs]:
```suggestion
            if (!rate.HasValue)
```

File: AccountingServer.BLL/Util/Exchange.cs [hunk 4]
```diff
@@ -150,4 +191,112 @@ private double Invoke(DateTime date, string from, string to)
...
+        /// <param name="to">目标</param>
+        /// <returns>汇率</returns>
+        private double Invoke(DateTime date, string from, string to)
+        {
+            if (from == to)
+                return 1;
+
+            var endpoint = date > DateTime.UtcNow
+                ? "latest"
+                : throw new ApplicationException("Historical rate unavailable.");
```
Comment (b1f6c1c4) [inline AccountingServer.BLL/Util/Exchange.cs]: https://coinmarketcap.com/api/documentation/v1/#operation/getV1ExchangeQuotesHistorical

**Review by b1f6c1c4 — CHANGES_REQUESTED** [review]

File: AccountingServer.BLL/Util/Exchange.cs (hunk 4 above)
Comment (b1f6c1c4) [reply AccountingServer.BLL/Util/Exchange.cs]: If it is paid plan we may just use latest rate...

**Comment by b1f6c1c4** [comment]
1. Call `/v1/cryptocurrency/map?symbol=...` to get CMC ID.
2. Cache CMC ID in memory.
3. Call `/v1/cryptocurrency/quotes/latest?id=...&convert=...` to get latest rate.

File: AccountingServer.BLL/Util/Exchange.cs (hunk 2 above)
Comment (IoriOikawa) [reply AccountingServer.BLL/Util/Exchange.cs]: done

File: AccountingServer.BLL/Util/Exchange.cs (hunk 4 above)
Comment (IoriOikawa) [reply AccountingServer.BLL/Util/Exchange.cs]: yep. we can stick on current rate then...
let's stick around alternatives, if possible in the future.

Comment (IoriOikawa) [reply AccountingServer.BLL/Util/Exchange.cs]: or, ANOTHER TAR PIT: a reinvented currency conversion part?

**Pull request closed without merging**